elements/pl-file-preview/pl-file-preview.js

```diff
@@ -0,0 +1,106 @@
+      throw new Error(`Failed to download file: ${result.status}`);
+    }
+    const blob = await result.blob();
+    const aElement = document.createElement('a');
+    aElement.setAttribute('download', name);
+    const href = URL.createObjectURL(blob);
+    aElement.href = href;
+    aElement.setAttribute('target', '_blank');
+    aElement.click();
+    URL.revokeObjectURL(href);
+  }
+
+  function escapePath(path) {
+    return path
+      .replace(/^\//, '')
+      .split('/')
+      .map((part) => encodeURIComponent(part))
+      .join('/');
+  }
+
+  class PLFilePreview {
+    constructor(uuid) {
+      const filePreview = document.querySelector('#file-preview-' + uuid);
+      const submissionFilesUrl = filePreview.dataset.submissionFilesUrl;
+
+      filePreview.querySelectorAll('li').forEach((li) => {
+        const file = li.dataset.file;
+        const escapedFileName = escapePath(file);
+        const path = `${submissionFilesUrl}/${escapedFileName}`;
+
+        const errorMessage = li.querySelector('.alert.error');
+
+        function showErrorMessage() {
+          errorMessage.classList.remove('d-none');
+        }
+
+        function hideErrorMessage() {
+          errorMessage.classList.add('d-none');
+        }
+
+        const downloadButton = li.querySelector('.file-preview-download-file');
+        downloadButton.addEventListener('click', (event) => {
+          event.stopPropagation();
+          downloadFile(path, file)
+            .then(() => {
+              hideErrorMessage();
+            })
+            .catch((err) => {
+              console.error(err);
+              showErrorMessage();
+            });
+        });
+
+        let wasOpened = false;
+        const preview = li.querySelector('.file-preview');
+        $(preview).on('show.bs.collapse', () => {
+          if (wasOpened) return;
+          const pre = preview.querySelector('pre');
+          const code = preview.querySelector('code');
+          const img = preview.querySelector('img');
+
+          fetch(path, { method: 'GET' })
+            .then((result) => {
+              if (!result.ok) {
+                throw new Error(`Failed to download file: ${result.status}`);
+              }
+              return result.blob();
+            })
+            .then(async (blob) => {
+              const type = blob.type;
+              if (type === 'text/plain') {
+                const text = await blob.text();
+                code.textContent = text;
+                pre.classList.remove('d-none');
+              } else if (type.startsWith('image/')) {
+                const url = URL.createObjectURL(blob);
+                img.src = url;
+                img.onload = () => {
+                  URL.revokeObjectURL(url);
+                };
+                img.classList.remove('d-none');
+              } else {
+                // We can't preview this file.
+                code.textContent = 'Content preview is not available for this type of file.';
```

Comment:
In this case can we write this text not inside the code element? For example, in the files below it looks like `foo.pdf` has rendered some content until you read the text. I'm thinking it might be better to just have plain text for the message.

<img width="788" alt="image" src="https://user-images.githubusercontent.com/875833/217309789-10a43f95-e05b-4ba8-838f-b2c887fcb7b6.png">

Reply:
Yep - I tweaked this in 19a115b. How does this look?

<img width="712" alt="Screenshot 2023-02-07 at 15 28 52" src="https://user-images.githubusercontent.com/1476544/217390277-e991cf75-a615-4523-a412-a44cd8964769.png">